hardware/plugins/PluginProtocols.cpp

```diff
@@ -883,169 +889,202 @@ namespace Plugins {
 
 	void CPluginProtocolMQTT::ProcessInbound(const ReadMessage * Message)
 	{
-		byte		bResponseType = Message->m_Buffer[0] & 0xF0;
-		PyObject*	pMqttDict = PyDict_New();
-		PyObject*	pObj = NULL;
-		long		iPacketIdentifier = 0;
-		long		iRemainingLength = Message->m_Buffer[1];
+		byte loop = 0;
+		m_sRetainedData.insert(m_sRetainedData.end(), Message->m_Buffer.begin(), Message->m_Buffer.end());
 
-		switch (bResponseType)
-		{
-		case MQTT_CONNACK:
-		{
-			ADD_STRING_TO_DICT(pMqttDict, "Verb", std::string("CONNACK"));
-			if (Message->m_Buffer[1] == 2) // check length is correct
+		do {
+			std::vector<byte>::iterator it = m_sRetainedData.begin();
+
+			byte		header = *it++;
+			byte		bResponseType = header & 0xF0;
+			PyObject*	pMqttDict = PyDict_New();
+			PyObject*	pObj = NULL;
+			long		iPacketIdentifier = 0;
+			long		iRemainingLength = 0;
+			long		multiplier = 1;
+			byte 		encodedByte;
+
+			do
 			{
-				ADD_INT_TO_DICT(pMqttDict, "Status", Message->m_Buffer[3]);
-				switch (Message->m_Buffer[3])
+				encodedByte = *it++;
+				iRemainingLength += (encodedByte & 127) * multiplier;
+				multiplier *= 128;
+				if (multiplier > 128*128*128)
 				{
-				case 0:
-					ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Connection Accepted"));
-					break;
-				case 1:
-					ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Connection Refused, unacceptable protocol version"));
-					break;
-				case 2:
-					ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Connection Refused, identifier rejected"));
-					break;
-				case 3:
-					ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Connection Refused, Server unavailable"));
-					break;
-				case 4:
-					ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Connection Refused, bad user name or password"));
-					break;
-				case 5:
-					ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Connection Refused, not authorized"));
-					break;
-				default:
-					ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Unknown status returned"));
-					break;
+					_log.Log(LOG_ERROR, "(%s) Malformed Remaining Length.", __func__);
+					return;
 				}
-			}
-			else
+			} while ((encodedByte & 128) != 0);
+
+			if (iRemainingLength > std::distance(it, m_sRetainedData.end()))
 			{
-				ADD_INT_TO_DICT(pMqttDict, "Status", -1);
-				ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Invalid message length"));
+				_log.Log(LOG_TRACE, "(%s) Too little data (got %u, expected %u).", __func__, std::distance(it, m_sRetainedData.end()), iRemainingLength);
```

Comment:
Is it possible to rephrase this to 'Not enough data received' instead of 'Too little data' ?

Reply:
Yes, sure!  Should I squash the commits as well?

Reply:
No need, i could smash it as we merge ;)